src/python/pants/engine/target_test.py

```diff
@@ -0,0 +1,191 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from collections import OrderedDict
+from dataclasses import dataclass
+from pathlib import PurePath
+from typing import ClassVar, List, Optional, Tuple
+
+import pytest
+
+from pants.engine.fs import EMPTY_DIRECTORY_DIGEST, PathGlobs, Snapshot
+from pants.engine.rules import UnionMembership, rule
+from pants.engine.selectors import Get
+from pants.engine.target import AsyncField, PluginField, PrimitiveField, Target
+from pants.testutil.engine.util import MockGet, run_rule
+from pants.util.collections import ensure_str_list
+from pants.util.memo import memoized_property
+from pants.util.ordered_set import OrderedSet
+
+
+class HaskellGhcExtensions(PrimitiveField):
+    alias: ClassVar = "ghc_extensions"
+    raw_value: Optional[List[str]]
+
+    @memoized_property
+    def value(self) -> List[str]:
+        if self.raw_value is None:
+            return []
+        # Add some arbitrary validation to test that hydration works properly.
+        bad_extensions = [
+            extension for extension in self.raw_value if not extension.startswith("Ghc")
+        ]
+        if bad_extensions:
+            raise ValueError(
+                f"All elements of `{self.alias}` must be prefixed by `Ghc`. Received "
+                f"{bad_extensions}."
+            )
+        return self.raw_value
+
+
+class HaskellSources(AsyncField):
+    alias: ClassVar = "sources"
+    raw_value: Optional[List[str]]
```

Comment:
How are default values modelled here? Is it just by adding ` = ["*.hs"]` to the end of this line?

Reply:
And as a follow-up, is there a way to set post-hydration defaults? i.e. is there a way for me to provide a default `Snapshot` rather than a default `PathGlobs`?

Reply:
Currently, default values would be applied in the corresponding rule `@hydrate_haskell_sources`.

Almost certainly, what will actually end up happening is that we create a `Sources(AsyncField)` class and have `PythonLibrarySources`, `PythonTestsSources`, etc all subclass it. That superclass would expect a field like `default_globs` and would use that to correctly apply them. We have total flexibility here because these are simple dataclasses that work with inheritance.

Reply:
> And as a follow-up, is there a way to set post-hydration defaults? i.e. is there a way for me to provide a default Snapshot rather than a default PathGlobs?

Not yet, but we could easily add that because these are simple Python dataclasses that respect inheritance. 

(For now, I'm trying to avoid adding any sugar until we know we actually need it.)